llvm/docs/Telemetry.rst

```diff
@@ -0,0 +1,246 @@
+      started = true;
+      object = std::make_unique<json::Object>();
+      return Error::success();
+    }
+
+    // Serialize the given value.
+    void write(StringRef KeyName, bool Value) override {
+      writeHelper(KeyName, Value);
+    }
+
+    void write(StringRef KeyName, int Value) override {
+      writeHelper(KeyName, Value);
+    }
+
+    void write(StringRef KeyName, size_t Value) override {
+      writeHelper(KeyName, Value);
+    }
+    void write(StringRef KeyName, StringRef Value) override {
+      writeHelper(KeyName, Value);
+    }
+
+    void write(StringRef KeyName,
+               const std::map<std::string, std::string>& Value) override {
+      json::Object Inner;
+      for (auto kv : Value) {
+        Inner.try_emplace(kv.first, kv.second);
+      }
+      writeHelper(KeyName, json::Value(std::move(Inner)));
+    }
+
+    Error finalize() override {
+      if (!started)
+        return createStringError("Serializer not currently in use");
+      started = false;
+      return Error::success();
+    }
+
+  private:
+    template <typename T> void writeHelper(StringRef Name, T Value) {
+      assert(started && "serializer not started");
+      object->try_emplace(Name, Value);
+    }
+    bool started = false;
+    std::unique_ptr<json::Object> object;
+  };
+       
+  class MyManager : public telemery::Manager {
+  public:
+  static std::unique_ptr<MyManager> createInstatnce(telemetry::Config* config) {
+    // If Telemetry is not enabled, then just return null;
+    if (!config->EnableTelemetry) return nullptr;
+
+    return std::make_unique<MyManager>();
+  }
+  MyManager() = default;
+
+  Error dispatch(TelemetryInfo* Entry) const override {
+    Entry->SessionId = SessionId;
+    emitToAllDestinations(Entry);
+  }
+      
+  void addDestination(std::unique_ptr<Destination> dest) override {
+    destinations.push_back(std::move(dest));
+  }
+  
+  // You can also define additional instrumentation points.
+  void logStartup(TelemetryInfo* Entry) {
+    // Add some additional data to entry.
+    Entry->Msg = "Some message";
+    dispatch(Entry);
+  }
+  
+  void logAdditionalPoint(TelemetryInfo* Entry) {
+    // .... code here
+  }
+  
+  private:
+    void emitToAllDestinations(const TelemetryInfo* Entry) {
+      for (Destination* Dest : Destinations) {
+        Dest->receiveEntry(Entry);
+      }
+    }
+    
+    std::vector<Destination> Destinations;
+    const std::string SessionId;
+  };
+
+  class MyDestination : public telemetry::Destination {
+  public:
+    Error receiveEntry(const TelemetryInfo* Entry) override {
+      if (Error err = serializer.init()) {
+        return err;
+      }
+      Entry->serialize(serializer);
+      if (Error err = serializer.finalize()) {
+        return err;
+      }
+
+      json::Object copied = *serializer.getOutputObject();
+      // Send the `copied` object to wherever.
+      return Error::success();
+    }
+
+  private:
+    JsonSerializer serializer;
```

Comment:
There are a few variables in this class that don't conform to LLVM naming style (i.e. they use lower case, not upper case).

Reply:
fixed (resolving)